src/NServiceBus.RavenDB/Subscriptions/SubscriptionPersister.cs

```diff
@@ -94,14 +99,42 @@ public SubscriptionPersister(IDocumentStore store)
         {
             var ids = messageTypes.Select(Subscription.FormatId).ToList();
 
-            using (var session = OpenAsyncSession())
+            using (var suppressTransaction = new TransactionScope(TransactionScopeOption.Suppress, TransactionScopeAsyncFlowOption.Enabled))
+            {
+                Subscriber[] subscribers;
+                using (var session = OpenAsyncSession())
+                {
+                    using (ConfigureAggressiveCaching(session))
+                    {
+                        var subscriptions = await session.LoadAsync<Subscription>(ids).ConfigureAwait(false);
+
+                        subscribers = subscriptions.Where(s => s != null)
+                            .SelectMany(s => s.Subscribers)
+                            .Distinct()
+                            .Select(c => new Subscriber(c.TransportAddress, c.Endpoint))
+                            .ToArray();
+                    }
+                }
+
+                suppressTransaction.Complete();
+                return subscribers;
+            }
+        }
+
+        IDisposable ConfigureAggressiveCaching(IAsyncDocumentSession session)
+        {
+            if (DisableAggressiveCaching)
             {
-                var subscriptions = await session.LoadAsync<Subscription>(ids).ConfigureAwait(false);
+                return new EmptyDisposable();
```

Comment:
if I recall correctly since the return type is IDisposable the struct will be boxed and therefore you end up having a heap allocation anyway. 

http://stackoverflow.com/questions/3032750/structs-interfaces-and-boxing

So in that case you can remove the struct. Make it a class and a static singleton which is way more efficient